token/token.go

```diff
@@ -447,15 +461,366 @@ func (mgm *tokenManager) GenerateServiceAccountToken(req *goa.RequestData, saID
+	expiresIn := keycloakToken.Extra("expires_in")
+	if expiresIn != nil {
+		extra["expires_in"] = expiresIn
+	}
+	refreshExpiresIn := keycloakToken.Extra("refresh_expires_in")
+	if refreshExpiresIn != nil {
+		extra["refresh_expires_in"] = refreshExpiresIn
+	}
+	notBeforePolicy := keycloakToken.Extra("not_before_policy")
+	if notBeforePolicy != nil {
+		extra["not_before_policy"] = notBeforePolicy
+	}
+	if len(extra) > 0 {
+		token = token.WithExtra(extra)
+	}
+
+	return token, nil
+}
+
+// GenerateUserTokenForIdentity generates an OAuth2 user token for the given identity
+func (mgm *tokenManager) GenerateUserTokenForIdentity(ctx context.Context, identity account.Identity) (*oauth2.Token, error) {
+	nowTime := time.Now().Unix()
+	unsignedAccessToken, err := mgm.GenerateUnsignedUserAccessTokenForIdentity(ctx, identity)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	accessToken, err := unsignedAccessToken.SignedString(mgm.userAccountPrivateKey.Key)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	unsignedRefreshToken, err := mgm.GenerateUnsignedUserRefreshTokenForIdentity(ctx, identity)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	refreshToken, err := unsignedRefreshToken.SignedString(mgm.userAccountPrivateKey.Key)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	var in30Days, nbf int64
+	in30Days = 30 * 24 * 60 * 60 // In 30 days
+
+	token := &oauth2.Token{
+		AccessToken:  accessToken,
+		RefreshToken: refreshToken,
+		Expiry:       time.Unix(nowTime+in30Days, 0),
+		TokenType:    "bearer",
+	}
+
+	// Derivative OAuth2 claims "expires_in" and "refresh_expires_in"
+	extra := make(map[string]interface{})
+	extra["expires_in"] = in30Days
+	extra["refresh_expires_in"] = in30Days
+	extra["not_before_policy"] = nbf
+
+	token = token.WithExtra(extra)
+
+	return token, nil
+}
+
+// GenerateUnsignedUserAccessToken generates an unsigned OAuth2 user access token for the given identity based on the Keycloak token
+func (mgm *tokenManager) GenerateUnsignedUserAccessToken(ctx context.Context, keycloakAccessToken string, identity *account.Identity) (*jwt.Token, error) {
+	token := jwt.New(jwt.SigningMethodRS256)
+	token.Header["kid"] = mgm.userAccountPrivateKey.KeyID
+
+	kcClaims, err := mgm.ParseToken(ctx, keycloakAccessToken)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	req := goa.ContextRequest(ctx)
+	if req == nil {
+		return nil, errors.New("missing request in context")
+	}
+
+	authOpenshiftIO := rest.AbsoluteURL(req, "", mgm.config)
+	openshiftIO, err := rest.ReplaceDomainPrefixInAbsoluteURL(req, "", "", mgm.config)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	claims := token.Claims.(jwt.MapClaims)
+	claims["jti"] = uuid.NewV4().String()
+	claims["exp"] = kcClaims.ExpiresAt
+	claims["nbf"] = kcClaims.NotBefore
+	claims["iat"] = kcClaims.IssuedAt
+	claims["iss"] = kcClaims.Issuer
+	claims["aud"] = kcClaims.Audience
+	claims["typ"] = "Bearer"
```

Comment:
Do we really need this ?
This is not present in the list https://www.iana.org/assignments/jwt/jwt.xhtml

Reply:
It's optional JWT claim which is used by KC. I think it's a good way to distinguish bearers from refresh tokens.